Update the structure of api_callable.

jmuk opened the pull request:
Now this allows getting call options as an optional parameter
and changes its behavior.

In a part of #104 effort.

geigerj commented on google/gax/__init__.py:
```diff
@@ -80,11 +74,21 @@ def __init__(self, timeout=30, retry=None, page_descriptor=None,
+    @property
+    def flatten_pages(self):
+        """
+        A boolean property whether a page streamed response should make
```
"property whether" --> "property indicating whether"

jmuk replied:
done

geigerj commented on google/gax/__init__.py:
```diff
@@ -80,11 +74,21 @@ def __init__(self, timeout=30, retry=None, page_descriptor=None,
+        """
+        A boolean property whether a page streamed response should make
+        the page structure transparent to the user by flattening the
+        repeated field in the returned generator.
```
s/generator/iterator

jmuk replied:
done

geigerj commented on google/gax/__init__.py:
```diff
@@ -103,8 +107,9 @@ def merge(self, options):
-                bundle_descriptor=self.bundle_descriptor)
+                page_descriptor=self.page_descriptor,
+                page_token=self.page_token,
+                bundler=self.bundler, bundle_descriptor=self.bundle_descriptor)
```
Good catch!

geigerj commented on google/gax/api_callable.py:
```diff
@@ -41,59 +41,63 @@
 
-def _add_timeout_arg(a_func, timeout):
-    """Updates a_func so that it gets called with the timeout as its final arg.
+class _RetryableWrapper(object):
```
@tbetbetbe made a past decision to move away from using callable objects to using our current functional/"decorator-like" style. What is the advantage in using an object here?

jmuk replied:
I could rewrite as decorator-style, but I think the code structure is mostly same and using callable class is much more straightforward in this case.

The point is that the decoration (the behavior) may change upon the invocation because it needs to change either of _add_timeout_arg or _retry, and they share some values like a_func and base settings. This is naturally a class, although we could implement it as internal functions.

geigerj replied:
I don't feel strongly about this, but I don't think this is particularly unnaturally expressed as a function either -- it seems to me that the only "object-y" thing about this is storing the read-only attribute `settings`, which works equally well in a closure.

I'm fine with it either way, although I would rather `settings` belonged to a closure (or callable object) in `create_api_call` than in the retrying part of the code.

geigerj commented on google/gax/api_callable.py:
```diff
@@ -41,59 +41,63 @@
+          options (gax.CallOptions): the optional value to customize the
+            calling behavior.
+        """
+        settings = self._settings.merge(options)
```
It seems uninuitive that the settings/options merge is happening inside an object called _RetryableWrapper, which I would expect to deal only with retrying. I would rather this happen in `create_api_call`.

jmuk replied:
`create_api_call` is not the case because, as is discussed in #104, `create_api_call` is called in the constructor like
```
self._create_shelf = gax.create_api_call(stub.create_shelf, defaults['create_shelf'])
```
and then the call options is passed in the individual method invocation, like:
```
def create_shelf(self, options):
   self._create_shelf(req, options, ...)
```

So the result of `create_api_call` should receive a CallOptions instance, and handling the overrides -- that is the role of merge method.

geigerj replied:
What about adopting the same structure for `create_api_call` as for the rest of the decorator-style functions?

```
def create_api_call(...)
  def inner(...)
    settings.merge(options)
    ...
```

?

jmuk replied:
I agree that settings.merge() will be in a single place, and restructured a bit. Please take another look. I am not sure this is the right direction though -- adding more and more inner functions of inner functions.

To be clear, this still determine _page_streamable/_bundleable/else when create_api_call() is invoked, for multiple reasons:
- this structure is actually determined when create_api_call() is called, so this is better and more understandable
- a test case actually require to raise an error for inconsistent bundling/page-streaming for create_api_call(), and I think that's the right thing to keep it.

geigerj commented:
Note: googleapis/toolkit#214 also changes the handling of `metadata`, which requires an update to `CallSettings` and `api_callable`.

jmuk commented:
I reviewed #104 discussion but unclear why metadata goes into construct_settings. It's actually not discussed well.

But I don't think that's the key part of the redesign, and therefore I think that's no problem to be in a part of method invocation, i.e.
```
self._create_shelf(request, options, metadata=self._headers)
```

And this structure would work well with this PR.

geigerj commented:
regarding metadata -- this was in Garrett's original mockup on the pubsub codegen, and is implicitly represented in #104 by the simplified method invocation in the second code block. If you feel that this should not be addressed in this PR, I can do it in a follow-up.

geigerj commented on google/gax/api_callable.py:
```diff
@@ -141,49 +142,48 @@ def _bundleable(a_func, desc, bundler):
 
-def _page_streamable(a_func, page_descriptor, page_token=None,
-                     flatten_pages=True):
+def _page_streamable(a_func, settings, page_descriptor):
```
The `page_descriptor` parameter is redundant with the `settings` parameter -- let's just use `settings`

geigerj replied:
whoops, was on the wrong commit

geigerj commented on google/gax/__init__.py:
```diff
@@ -429,3 +431,33 @@ def next(self):
+    def next(self):
+        """Retrieves the next resource."""
+        # pylint: disable=next-method-called
+        while not self._current:
```
Why the while loop? That is, when will a PageIterator contain a falsy value?

jmuk replied:
PageIterator could return an empty array but have a next page token, in that case this should skip the empty array and fetch the next page.

jmuk commented:
Regarding metadata -- rethinking about that, I'm okay with merging into construct_settings.
But I'm thinking `CallSettings` might not be a good name now -- it keeps the option parameters (like backoff settings) and the information of the API-call such as page descriptions, bundle descriptions, and now metadata (headers) -- those are properties of the method itself rather than invocations (= call).
This could be `MethodSettings`?

geigerj commented:
SGTM regarding `CallSettings` vs `MethodSettings`. I think there has always been some ambiguity from the naming of `CallOptions` and `CallSettings` about the difference -- hopefully this name change will make it clearer.

jmuk commented:
Updated about the metadata -- not renamed the class yet. I think it can be done in another PR.

Also, I think GAX should not be strongly tied to gRPC itself -- it should be encapsulated into grpc.py file only as far as possible, and therefore, I've added 'kwargs' argument instead.

geigerj commented on google/gax/__init__.py:
```diff
@@ -134,7 +139,8 @@ def merge(self, options):
                 page_descriptor=self.page_descriptor, page_token=page_token,
-                bundler=bundler, bundle_descriptor=self.bundle_descriptor)
+                bundler=bundler, bundle_descriptor=self.bundle_descriptor,
+                kwargs=self.kwargs)
```
Let's a actually do

```
if options.kwargs == OPTION_INHERIT:
  kwargs = self.kwargs
else:
  kwargs = self.kwargs.copy()
  kwargs.update(options.kwargs)
```

I can't tell where `options.kwargs` is currently being used -- maybe we are just ignoring it at the moment?

jmuk replied:
added -- no such options.kwargs exist currently, so I've added it.

geigerj commented on google/gax/api_callable.py:
```diff
@@ -88,7 +88,7 @@ def _retryable(a_func, retry):
     total_timeout = (retry.backoff_settings.total_timeout_millis /
                      _MILLIS_PER_SECOND)
 
-    def inner(*args, **kwargs):
```
This is a good cleanup -- previously, both CallOptions and the inner call took `kwargs`, which was redundant. This is better.

geigerj replied:
Optional: we should ensure that the "decorator"/transformation functions all look like this, i.e., the final call's `kwargs` comes only from Call{Settings,Options} and not from actual function parameters. This can also be done in a follow-up PR.

jmuk replied:
agreed, let me do it in another PR

geigerj commented:
Just one concern about the CallSettings `merge` method. Otherwise looks good

geigerj commented:
LGTM.

I will update googleapis/toolkit#214 to put metadata into kwargs.
